_posts/2023-12-06-JDBC-sink-connector-batch-support.adoc

```diff
@@ -0,0 +1,160 @@
+
+*CPU*: Intel(R) Core(TM) i7-1185G7 @ 3.00GHz (8 cores) +
+*RAM*: 32GB +
+*Disk*: 512GB NVMe +
+
+All required components (Kafka, Connect, Zookeeper, etc.) inside docker containers.
+
+The table used for the test has the following structure:
+
+```sql
+CREATE TABLE `aviation` (
+  `id` int NOT NULL,
+  `aircraft` longtext,
+  `airline` longtext,
+  `passengers` int DEFAULT NULL,
+  `airport` longtext,
+  `flight` longtext,
+  `metar` longtext,
+  `flight_distance` double DEFAULT NULL
+)
+```
+
+=== Test plan
+We planned to execute these tests:
+
+* 100K events from single table
+** MySQL batch vs without batch
+* 100K events from three different table
+** MySQL batch vs without batch
+* 1M events from single table
+** MySQL batch with batch size: 500, 1000, 5000, 10000 vs without batch
+** MySQL batch with batch size: 500, 1000, 5000, 10000 with JSONConverter
+** MySQL batch with batch size: 500, 1000, 5000, 10000 with Avro
+** MySQL batch with batch size: 500, 1000, 5000, 10000 with Avro and no index on destination table
+
+:imagesdir: /assets/images/2023-12-06-JDBC-sink-connector-batch-support
+
+.{nbsp}
+image::100k-batch-no-batch.png[role=centered-image]
+
+_Figure 1_ illustrates the total execution time required to process 100,000 events from a single table, comparing MySQL connector with and without the batch support.
+
+[NOTE]
+====
+Despite the default values being set to `500` for both `batch.size` and `consumer.max.poll.records`, the observed actual size was reduced to `337` records due to payload size considerations.
+====
+
+We can observe, as expected, that the Debezium JDBC connector with batch support is faster.
+
+.{nbsp}
+image::100k-3-tables.png[role=centered-image]
+
+In _Figure 2_, we observe that splitting 100,000 events into three tables does not impact the results.
+The Debezium JDBC connector with batch support remains faster compared to the non-batch version.
+
+.{nbsp}
+image::1M-batch-no-batch.png[role=centered-image]
+
+In _Figure 3_, it is evident that the performance gain becomes more pronounced with 1,000,000 events.
+The Debezium JDBC connector with batch support took approximately 7 minutes to insert all events, with an average throughput of 2300 eps, while the process without batch support took 570 minutes (9.5 hours).
+Therefore, the Debezium JDBC connector with batch support is 79 times faster than the version without batch support.
+
+.{nbsp}
+image::1M-different-batch-size-json.png[role=centered-image]
+
+In _Figure 4_, we observe the behavior of the Debezium JDBC connector writing on MySQL with different batch.size.
+Although the difference is notable at the beginning, it becomes apparent that the throughput continues to slow down, and, on average,
+all the `batch.size` configurations take about 7 minutes to process all events.
+
+This raised a concern for us. After conducting a thorough analysis (profiling), we identified that the bottleneck was in event deserialization.
```

Comment:
Does it really explains the difference between batches at the beginning? If deserialization is the root cause here, I'd expect that there won't be any difference also at the beginning. This looks like there are possibly also other factors which may affect it (my tip would be some kind of buffering somewhere).

If you have any explanation for it, I'd add it here.

Reply:
Good point. Maybe I think I should better clarify that this strange behavior lead us to investigate and we have found the problem with ser/der that solved the an issue but not the slow down that can be related to other stuff that we have not investigated.

Reply:
yes, making this more clear would IMHO make the this part more understandable

Reply:
@vjuranek  Do you think it's more clear now?